src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -14,68 +13,42 @@
-            s_removeUnreadMembersRule = CreateDescriptor(IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId,
-                                                         removeUnreadMembersTitle,
-                                                         removeUnreadMembersMessage,
-                                                         configurable: true,
-                                                         enabledByDefault: true);
-            s_removeUnreadMembersWithFadingRule = CreateUnnecessaryDescriptor(IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId,
-                                                                              removeUnreadMembersTitle,
-                                                                              removeUnreadMembersMessage,
-                                                                              configurable: true,
-                                                                              enabledByDefault: true);
-        }
+        private static DiagnosticDescriptor CreateDescriptor(string id, LocalizableResourceString title, LocalizableResourceString messageFormat)
+            => new DiagnosticDescriptor(id, title, messageFormat, DiagnosticCategory.CodeQuality,
+                    DiagnosticSeverity.Info, isEnabledByDefault: true, customTags: DiagnosticCustomTags.Unnecessary);
 
-        protected AbstractRemoveUnusedMembersDiagnosticAnalyzer()
-            : base (ImmutableArray.Create(s_removeUnusedMembersRule,
-                                          s_removeUnusedMembersWithFadingRule,
-                                          s_removeUnreadMembersRule,
-                                          s_removeUnreadMembersWithFadingRule))
-        {
-        }
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_removeUnusedMembersRule, s_removeUnreadMembersRule);
 
-        public override bool OpenFileOnly(Workspace workspace) => false;
+        public bool OpenFileOnly(Workspace workspace) => false;
 
         // We need to analyze the whole document even for edits within a method body,
         // because we might add or remove references to members in executable code.
         // For example, if we had an unused field with no references, then editing any single method body
         // to reference this field should clear the unused field diagnostic.
         // Hence, we need to re-analyze the declarations in the whole file for any edits within the document. 
-        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
```

Comment:
I think this category is just worth to be public. I can either in DiagnosticAnalyzer or we have some ISupportIDEDiagnosticAnalyzer or something that if implemented, it can be more optimized. and we could use such an interface to make something more efficient in IDE like compilation end code fix or something...

Reply:
if we expose the category public and make OpenFileOnly as a tag or exposed it as just API, we can remove IBuiltInAnalyzer

Reply:
> if we expose the category public 

I am hoping we can do even better and automatically derive the category based on the set of registered actions, and also allow explicit override from analyzer through custom tags. An analyzer should ideally honor the implicit contract of analyzing within the code unit associated with the callback context. If not, that is actually a bug in analyzer's chosen registration approach and a very good indication of a performance issue. For example, if an analyzer registers for  invocation syntax nodes, but ends up analyzing the whole method body or across methods/types, it is a poorly written analyzer with certain performance problems (ideally should register a compilation start action or symbol start action to match its analysis scope). For well written analyzers, with ideal performance characteristics, action registration should directly imply the analyzer category.

Reply:
that's kind of true if make things a bit more restrictive. but there were arguments from many people saying analyzer can report diagnostics anywhere regardless of where it is analyzing.

basically, there are 2 scopes. 1 that being analyzed (1). and 1 that being reported (2). 

diagnostic analyzer uses (1) to produce (2). and code fix uses (2) to offer a fix.

problem is code fix can't figure out (1) from (2). the category gives a hint to code fix how to figure out (1) from (2)

unless, we will start to restrict where one can report diagnostics (2) related to (1). it would be hard to induce (1) from (2)

here.. (1) is scope analyzer is analyzing. (2) is the scope where analyzer reports diagnostics. for example, analyzing a method (M) in symbol (A) reporting diagnostics on a property (P) in symbol (B)

M is (1) and P is (2)